README.md
# snappy-dev-website

This repo contains tools and generation documentation for the backend generations on snappy developer website.

Running `generate` will create an *out/* directory, containing the documentation for any supported releases and devices,
ready to be imported into our developer site. This one also import remote branch documentation from snappy, snapcraft and
demo branches.


releases
# This is the release-git branch format mapping
# Format is:
# RELEASE GIT_BRANCH_NAME
16.04 master
#15.04 15.04


sitegenerator/tools.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Copyright (C) 2016 Canonical
#
# Authors:
#  Didier Roche
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; version 3.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

'''General tools for handling files and metadata'''

import argparse
from contextlib import contextmanager
import logging
import os
import sys

from .settings import OUTPUT_DIR

logger = logging.getLogger(__name__)


def next_relevant_line(f):
    '''Return stripped relevant next line. Ignore comments and blank ones'''
    for line in f:
        if not line or not line.strip() or line.startswith("#"):
            continue
        yield line.strip()


@contextmanager
def replace_file_inline(path):
    '''Replace a file atomically creating a temp .new one

    Skip non text files'''

    temp_file = "{}.new".format(path)
    try:
        with open(temp_file, 'w') as dest_f:
            with open(path) as source_f:
                yield (source_f, dest_f)
        os.rename(temp_file, path)
    except UnicodeDecodeError as e:
        # That should be the case of any binary files
        logger.debug("Couldn't replace in {}: {}".format(path, e))
        os.remove(temp_file)


def setup_args():
    '''Handle CLI arg setup'''
    parser = argparse.ArgumentParser(description="Generate in the {} directory the developer website "
                                                 "content, fetching from different sources.".format(OUTPUT_DIR))
    parser.add_argument("-v", "--verbose", action="count", default=0, help="Increase output verbosity (2 levels)")
    parser.add_argument("-d", "--debug", action="store_true", help="Max verbose output (2nd level)")

    args = parser.parse_args(sys.argv[1:])
    setup_logging_level(args)


def setup_logging_level(args):
    level = logging.WARNING
    if args.verbose == 1:
        level = logging.INFO
    if args.verbose > 1 or args.debug:
        level = logging.DEBUG

    logging.basicConfig(level=level, format="%(levelname)s: %(message)s")


sitegenerator/releases.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Copyright (C) 2016 Canonical
#
# Authors:
#  Didier Roche
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; version 3.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

'''Release-related misc utilities'''

import logging
import os
import sys
import subprocess
import yaml

from .settings import ROOT_DIR, RELEASES_BRANCH_MAPPING, VARIABLES_MAPPING
from .tools import next_relevant_line

logger = logging.getLogger(__name__)


def _get_supported_releases_map():
    '''Load supported release and master branch map'''
    release_map = {}
    with open(os.path.join(ROOT_DIR, RELEASES_BRANCH_MAPPING)) as f:
        try:
            for line in next_relevant_line(f):
                (release, branch_name) = line.split(" ")
                release_map[release] = branch_name
        except ValueError:
            logger.error("Release file is not of valid format: <release_name> <branch>")
            sys.exit(1)
    return release_map


def get_releases_in_context():
    '''Prepare for one release, switching context and return release name'''
    for release, branch in _get_supported_releases_map().items():
        subprocess.check_call(["git", "checkout", "-q", branch])
        yield release


def load_device_metadata(release):
    """Return maps of variables substitution token for each device.

    Adding RELEASE_VERSION and IMAGE_FILENAME (if IMAGE_URL present) automatically to each
    device for convenience.

    Format is:
        { 'device-key':
            { 'VARIABLE_NAME': 'VALUE' },
              …
            },
          …
        }

        Example:
        { 'rpi2': {'IMAGE_URL': 'https://download.ubuntu.com/blablabnla/rpi2-16.04.iso',
                   'IMAGE_FILENAME': 'rpi2-16.04.iso',
                   'RELEASE_VERSION': '16.04' },
          'dragonboard': {'FOO': 'BAR',
                          'RELEAS_VERSION': '16.04' },
        }"""
    devices_metadata = {}
    with open(os.path.join(ROOT_DIR, VARIABLES_MAPPING)) as f:
        devices_metadata = yaml.load(f.read())
    for device_key in devices_metadata:
        devices_metadata[device_key]['RELEASE_VERSION'] = release
        image_url = devices_metadata[device_key].get('IMAGE_URL')
        if image_url:
            devices_metadata[device_key]['IMAGE_FILENAME'] = os.path.basename(image_url)
    return devices_metadata


sitegenerator/settings.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Copyright (C) 2016 Canonical
#
# Authors:
#  Didier Roche
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; version 3.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

'''Variables and settings constants'''

import os

ROOT_DIR = os.path.dirname(os.path.dirname(__file__))
SITE_SRC = os.path.join(ROOT_DIR, "src")

OUTPUT_DIR = os.path.join(ROOT_DIR, "out")

# important mapping files
RELEASES_BRANCH_MAPPING = "releases"
GIT_IMPORT_MAPPING = "import-mapping"
VARIABLES_MAPPING = "variables-mapping"
IMPORT_URL_NAME = ".import_url"

PREPEND_TOUR_TEMPLATES = {
    "as-dev": "This content is part of the snappy developer tour, feel free to hop on it!",
    # "as-devops": "This content is part of the snappy devops tour, feel free to hop on it!",
    # "as-boardmaker": "This content is part of the snappy board maker tour, feel free to hop on it!"
}


sitegenerator/gitimporter.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Copyright (C) 2016 Canonical
#
# Authors:
#  Didier Roche
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; version 3.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

'''Handling import from different external git branches'''

import logging
import os
import shutil
import subprocess

from .tools import next_relevant_line
from .settings import ROOT_DIR, GIT_IMPORT_MAPPING, IMPORT_URL_NAME

logger = logging.getLogger(__name__)


def import_git_external_branches(out_root_dir, temp_repos_dir, release):
    '''Import and create needed path for each external git repository'''

    success = True
    with open(os.path.join(ROOT_DIR, GIT_IMPORT_MAPPING)) as f:
        try:
            for line in next_relevant_line(f):
                (rel_dest_path, git_repo, branch_name, copy_path) = line.split(" ")

                # create destination path, changing release with existing release
                rel_dest_path = rel_dest_path.replace("release", release)
                dest_path = os.path.join(out_root_dir, rel_dest_path)
                os.makedirs(os.path.dirname(dest_path), exist_ok=True)

                repo_mirror = os.path.join(temp_repos_dir, git_repo.split("/")[-1])
                # ensure we have the repo in temp_repos_dir, otherwise create it
                if not os.path.isdir(repo_mirror):
                    subprocess.check_call(["git", "clone", "-q", "-n", git_repo], cwd=os.path.dirname(repo_mirror))
                subprocess.check_call(["git", "checkout", "-q", branch_name], cwd=repo_mirror)
                shutil.copytree(os.path.join(repo_mirror, copy_path), dest_path, ignore=lambda src, names: [".git"])
                open(os.path.join(dest_path, IMPORT_URL_NAME), "w").write("{}/tree/{}/{}".format(git_repo, branch_name,
                                                                                                 copy_path))
        except ValueError:
            logger.error("Import file is not of valid format: <site_path> <repo_url> <branch_name> "
                         "<path_to_content_copy_in_branch>")
            success = False
    return success


def find_imported_branch_metadata(path, root_dir):
    '''Find if the current path is an imported branch, and if so, get the link to it'''
    cur_path = path
    while (os.path.abspath(cur_path) != os.path.dirname(root_dir) and
            os.path.abspath(cur_path) != os.path.dirname(cur_path)):
        try:
            branch_link = open(os.path.join(cur_path, IMPORT_URL_NAME)).read()
            return os.path.normpath(os.path.join(branch_link, os.path.relpath(path, cur_path)))
        except FileNotFoundError:
            pass
        cur_path = os.path.dirname(path)

    return None


sitegenerator/fileshandling.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Copyright (C) 2016 Canonical
#
# Authors:
#  Didier Roche
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; version 3.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

'''Copying and treating files internal metadata'''

import logging
import os
import re
import shutil

from .tools import replace_file_inline

logger = logging.getLogger(__name__)

import_re = re.compile("##IMPORT (.*)")

required_variable_re = re.compile('\[\[([^\[\]]*)\]\]')
optional_variable_re = re.compile('\<\<([^<>]*)\>\>')

relative_markdown_links = re.compile("\(((?!http|www).[^\)]*\.md)\)")


def import_and_copy_file(source_path, destination_path):
    '''Copy and import file content.

    We handle:
        1. symlinks are replaced with real files
        2. ##IMPORT <file_path> to copy destination file content into current file

    We return an error if we couldn't copy or import all listed filed
    '''
    success = True
    try:
        with open(destination_path, 'w') as dest_f:
            with open(source_path) as source_f:
                success = _copycontent_withimports_tag(source_f, dest_f)
    except UnicodeDecodeError as e:
        # Fall back to direct copy for binary files
        logger.debug("Directly copy as can't read {} as text: {}".format(source_path, e))
        shutil.copy2(source_path, destination_path)
    return success


def _copycontent_withimports_tag(source_f, dest_f):
    '''Copy content from one file to the other, handling import tags

    imports within imports are handled, as well as symlinks chaining'''
    success = True
    for line in source_f:
        result = import_re.findall(line)
        if result:
            rel_path = result[0]
            try:
                with open(os.path.join(os.path.dirname(os.path.realpath(source_f.name)), rel_path)) as import_f:
                    if not _copycontent_withimports_tag(import_f, dest_f):
                        success = False
            except FileNotFoundError:
                logger.error("Couldn't import {} from {}".format(rel_path, source_f.name))
                success = False
        else:
            dest_f.write(line)
    return success


def _replace_from_map(line, regexp, replace_pattern, device_vars):
    '''Abstract replacing the variable from map.

    returning tuple is: (newline, failed_keywords_list)
    '''
    unfound_keywords_list = []
    for keyword in regexp.findall(line):
        try:
            line = line.replace(replace_pattern.format(keyword), device_vars[keyword])
        except KeyError:
            unfound_keywords_list.append(keyword)
    return (line, unfound_keywords_list)


def _replace_line_content(line, filename, device_name, device_vars):
    '''Return current line with replaced variable substitution.

    returning tuple is: (newline, success)
    '''
    success = True

    # handle optional variables first
    replace_pattern = "<<{}>>"
    (line, unfound_keywords) = _replace_from_map(line, optional_variable_re, replace_pattern, device_vars)
    for keyword in unfound_keywords:
        logger.info("{} doesn't have any mapping for {} which is optional in {}".format(
                device_name, keyword, filename))

    # handle required variables
    replace_pattern = "[[{}]]"
    (line, unfound_keywords) = _replace_from_map(line, required_variable_re, replace_pattern, device_vars)
    for keyword in unfound_keywords:
        logger.error("{} doesn't have any mapping for {} which is required in {}".format(
                device_name, keyword, filename))
        success = False

    return (line, success)


def replace_variables(path, device_name=None, device_vars={}):
    '''This variable replacement is done on files being in a per device directory.

    We handle variable substitution (only if device_name is provided)
        [[VARIABLE]] are required variables. It will print an error and return as such (not interrupting though)
        <<VARIABLE>> are optional variables. It will print an info and not return an error
    '''
    success = True
    with replace_file_inline(path) as (source_f, dest_f):
        for line in source_f:
            (line, new_success) = _replace_line_content(line, path, device_name, device_vars)
            success = new_success and success
            dest_f.write(line)

    return success


def reformat_links(path):
    '''Strip down the final .md on any relative path in links as it will be replaced with real file names'''
    with replace_file_inline(path) as (source_f, dest_f):
        for line in source_f:
            for link_to_replace in relative_markdown_links.findall(line):
                line = line.replace(link_to_replace, link_to_replace[:-3])
            dest_f.write(line)


def prepend_external_link(path, message, url):
    '''Prepend to existing page some known template type'''

    with replace_file_inline(path) as (source_f, dest_f):
        line_count = 1
        for line in source_f:
            # add the message as a second line
            if line_count == 2:
                dest_f.write("> [{}]({})\n".format(message, url))
            line_count += 1
            dest_f.write(line)


sitegenerator/__init__.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Copyright (C) 2016 Canonical
#
# Authors:
#  Didier Roche
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; version 3.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

import glob
import logging
import os
import shutil
import sys
import tempfile

from . import settings
from .fileshandling import import_and_copy_file, prepend_external_link, replace_variables, reformat_links
from .gitimporter import import_git_external_branches, find_imported_branch_metadata
from .releases import get_releases_in_context, load_device_metadata

logger = logging.getLogger(__name__)


def main():
    '''Main entry point'''
    logging.basicConfig(level=logging.DEBUG, format="%(levelname)s: %(message)s")

    success = True

    # Ensure that the generated directory doen't exist. If it exists, bails out
    if os.path.exists(settings.OUTPUT_DIR):
        logger.error("{} exists, please delete it first".format(settings.OUTPUT_DIR))
        sys.exit(1)
    os.makedirs(settings.OUTPUT_DIR)

    if not os.path.exists(settings.SITE_SRC):
        logger.error("{} is required".format(settings.SITE_SRC))
        sys.exit(1)

    # Handling unversioned doc from current branch
    unversioned_src_dir = os.path.join(settings.SITE_SRC, "unversioned")
    for path, dirs, files in os.walk(unversioned_src_dir):
        for file_name in files:
            file_path = os.path.join(path, file_name)
            dest_path = file_path.replace(unversioned_src_dir, settings.OUTPUT_DIR)
            os.makedirs(os.path.dirname(dest_path), exist_ok=True)
            if not import_and_copy_file(file_path, dest_path):
                success = False
            reformat_links(dest_path)

    # Loop and switch for each release context
    for release in get_releases_in_context():
        versioned_src_dir = os.path.join(settings.SITE_SRC, "versioned")

        # Load device metadata and variable substitution
        devices = load_device_metadata(release)

        # 1. Handle external branch imports (because other pages can import them)
        with tempfile.TemporaryDirectory() as tmp_dirname:
            if not import_git_external_branches(settings.OUTPUT_DIR, tmp_dirname, release):
                success = False

        # 2. Handle local directory and files copy + import statements in files
        for path, dirs, files in os.walk(versioned_src_dir):
            for file_name in files:
                file_path = os.path.join(path, file_name)
                # Replace the "release" keyword with the current release
                dest_path = file_path.replace(versioned_src_dir, settings.OUTPUT_DIR).replace("release", release)
                os.makedirs(os.path.dirname(dest_path), exist_ok=True)
                if not import_and_copy_file(file_path, dest_path):
                    success = False

        # 3. Do device variables and links replacement
        for path, dirs, files in os.walk(settings.OUTPUT_DIR):
            # Only process files in a path corresponding to this release
            if not "/{}".format(release) in path:
                continue
            for file_name in files:
                file_path = os.path.join(path, file_name)
                device_path_candidate = path.split("/")[-1]
                if device_path_candidate in devices:
                    if not replace_variables(file_path, device_path_candidate, devices[device_path_candidate]):
                        success = False
                reformat_links(file_path)

        # 4. Handle get-started and prepend "this is part of the tour" link
        for device_path in glob.glob(os.path.join(settings.OUTPUT_DIR, "guides-and-reference", release, "setup", "*")):
            device = device_path.split("/")[-1]
            for file in os.listdir(device_path):
                src_path = os.path.join(device_path, file)
                dest_file_name = "step2-setup-{}-{}".format(device, file)
                for tour_type in settings.PREPEND_TOUR_TEMPLATES:
                    tour_base_path = os.path.join(settings.OUTPUT_DIR, "get-started", tour_type)
                    dest_path = os.path.join(tour_base_path, release, dest_file_name)
                    relative_link = os.path.relpath(tour_base_path, os.path.dirname(src_path))
                    shutil.copy2(src_path, dest_path)
                    prepend_external_link(src_path, settings.PREPEND_TOUR_TEMPLATES[tour_type], relative_link)

        # 5. Rename landing pages (for examples, demos, codelabs) and stick URL if anything found
        for path, dirs, files in os.walk(settings.OUTPUT_DIR):
            # Only process files in a path corresponding to this release
            if not "/{}".format(release) in path:
                continue
            if "index.md" in files or "index.html" in files:
                continue
            if "README.md" in files:
                dest_file = os.path.join(path, "index.md")
                os.rename(os.path.join(path, "README.md"), dest_file)
                url = find_imported_branch_metadata(path, settings.OUTPUT_DIR)
                if url:
                    prepend_external_link(dest_file, "This content is available for checkout", url)

    if not success:
        logger.error("The site generation returned an error")
        sys.exit(1)
